clouddriver-google-common/src/main/groovy/com/netflix/spinnaker/clouddriver/googlecommon/deploy/GoogleCommonSafeRetry.java

```diff
@@ -0,0 +1,203 @@
+  }
+
+  /**
+   * Retry an operation if it fails. Treat any error codes in successCodes as success.
+   *
+   * @param operation - The operation.
+   * @param description - Description of the operation, used for logging.
+   * @param retryCodes - GoogleJsonResponseException codes we retry on.
+   * @param successCodes - GoogleJsonException codes we treat as success.
+   * @return Object returned from the operation.
+   */
+  @Nullable
+  public <V> V doRetry(
+      Callable<V> operation,
+      String description,
+      List<Integer> retryCodes,
+      List<Integer> successCodes,
+      Map<String, String> tags,
+      Registry registry) {
+    boolean success = false;
+    long startTime = registry.clock().monotonicTime();
+    try {
+      V result = performOperation(operation, description, retryCodes, successCodes);
+      success = true;
+      return result;
+    } catch (GoogleJsonResponseException e) {
+      throw GoogleApiException.fromGoogleJsonException(e);
+    } catch (SocketTimeoutException e) {
+      throw new GoogleApiException("Operation failed. Last attempt timed out.");
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new GoogleApiException("Operation failed. Thread was interrupted waiting to retry.");
+    } catch (Exception e) {
+      throw new IllegalStateException("Operation failed.", e);
+    } finally {
+      Map<String, String> metricTags =
+          ImmutableMap.<String, String>builder()
+              .putAll(tags)
+              .put("success", Boolean.toString(success))
+              .build();
+      registry
+          .timer(registry.createId("google.safeRetry", metricTags))
+          .record(registry.clock().monotonicTime() - startTime, TimeUnit.NANOSECONDS);
+    }
+  }
+
+  @Nullable
+  private <V> V performOperation(
+      Callable<V> operation,
+      String description,
+      List<Integer> retryCodes,
+      List<Integer> successfulErrorCodes)
+      throws Exception {
+    long maxAttempts = Math.max(1, maxRetries);
+    int tries = 1;
+    // This logic runs maxAttempts - 1 times, as we don't catch exceptions on the last try
+    while (tries < maxAttempts) {
+      try {
+        return attemptOperation(operation, successfulErrorCodes);
+      } catch (GoogleJsonResponseException jsonException) {
+        if (!retryCodes.contains(jsonException.getStatusCode())) {
+          throw jsonException;
+        }
+        log.warn(
+            "{} attempt #{} encountered retryable statusCode={} with error message: {}.",
+            description,
+            tries,
+            jsonException.getStatusCode(),
+            jsonException.getMessage());
+      } catch (SocketTimeoutException toEx) {
+        log.warn("Retryable {} attempt #{} timed out.", description, tries);
+      }
+      tries++;
+      // Sleep with exponential backoff based on the number of retries. Add retry jitter with
```

Comment:
Wow, I had never heard the term "retry jitter" to describe this special case of exponential backoff...love it.
♻️ 🕺 🐞